File: Python/_109_Convert_Sorted_List_to_Binary_Search_Tree/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        tree = []
        while head:
            tree.append(head.val)
            head = head.next
        root = self.sortedArrayToBST(tree)        
        return root
    
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if len(nums) <= 0:
            return None
        else:
            root = TreeNode(nums[len(nums) // 2])
            root.left = self.sortedArrayToBST(nums[:(len(nums) // 2)])
            root.right = self.sortedArrayToBST(nums[(len(nums) // 2 + 1):])
            return root

File: Java/_109_Convert_Sorted_List_to_Binary_Search_Tree/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        int size = 0;
        ListNode temp = head;
        while (temp != null) {
            ++size;
            temp = temp.next;
        }
        int[] tree = new int[size];
        int i = 0;
        while (head != null) {
            tree[i] = head.val;
            head = head.next;
            ++i;
        }
        TreeNode root = sortedArrayToBST(tree);     
        return root;
    }
    
    public TreeNode sortedArrayToBST(int[] nums) {
        if (nums.length <= 0)
			return null;
		else {
			TreeNode root = new TreeNode(nums[nums.length / 2]);
			int[] left_nums = Arrays.copyOfRange(nums, 0, nums.length / 2);
			int[] right_nums = Arrays.copyOfRange(nums, nums.length / 2 + 1, nums.length);
			root.left = sortedArrayToBST(left_nums);
			root.right = sortedArrayToBST(right_nums);
			return root;
		}
    }
}